docs/Foundation/NSInvocation.md

```diff
@@ -0,0 +1,127 @@
+
+Herein lies the challenge: **NSInvocation** must be able to lay out arguments at runtime, based solely on the Objective-C method signature
+provided during its construction.
+
+<sup>1. Procedure Call Standard for the ARM Architecture. [PDF link](http://infocenter.arm.com/help/topic/com.arm.doc.ihi0042f/IHI0042F_aapcs.pdf)</sup>
+
+## Architecture Decisions
+
+NSInvocation will rely on an internal platform-specific interface, `_NSInvocationCallFrame`, to do its argument allocation.
+
+`_NSInvocationCallFrame` has only three responsibilities: determining whether a method requires special struct return handling, storing an argument, and loading an argument.
+
+Therefore, platform support for NSInvocation will be provided via two discrete modules:
+
+1. A conformant implementation of `_NSInvocationCallFrame`.
+2. An assembly thunk that will take the platform internal representation of `_NSInvocationCallFrame` and dispatch a call to a function pointer with the stored arguments.
+
+### X86 details
+
+X86 is the easier of the two currently-supported architectures for which to allocate arguments.
+
+* Every argument is placed on the stack, in stack-ascending order.
+* Arguments less than four bytes in width are sign- or zero-extended and stored in one doubleword and emitted onto the stack.
+* Arguments greater than four bytes in width are promoted to an even number of doublewords and emitted onto the stack.
+* Values are returned in `eax` (if less than or equal to four bytes), `eax+edx` (if between four and eight bytes) and via hidden struct parameter.
+
+### ARM details
+
+Arguments in ARM are split between general-purpose registers, floating point registers, and the stack based on a few rules. While the rules are laid out in the *Procedure Call Standard*,
+they will be included here in brief.
+
+* Non-floating-point arguments less than or equal to four bytes in width are sign- or zero-extended and stored preferentially in registers `R0` through `R3`.
+  * If there are no unallocated registers, these arguments are placed on the stack.
+* Non-floating-point arguments eight bytes in width requiring 8-byte alignment are stored preferentially in `R0,R1` or `R2,R3`.
+  * If these registers are not available, the entire argument is placed on the stack.
+* Single-precision floating point arguments are preferentially allocated to single-precision registers `S0` through `S15`.
+  * If there are no unallocated single-precision registers, these arguments are placed on the stack.
+* Double-precision floating point arguments are preferentially allocated to double-precision registers `D0` through `D7`.
+  * If there are no unallocated double-precision registers, these arguments are aligned to 8 bytes and on the stack.
+  * Each `Dx` register consumes two `Sx` registers. They will be unavailable for floating-point arguments.
+* Aggregates of 1-4 single- or double-precision values are allocated to available floating-point registers.
```

Comment:
For what it's worth, objc does not have a type encoding for vector types, IIRC

---
In reply to: [76461957](https://github.com/Microsoft/WinObjC/pull/903#discussion_r76461957) [](ancestors = 76461957,76460844)